lib/FrontendTool/ReferenceDependencies.cpp

```diff
@@ -402,24 +395,138 @@ bool swift::emitReferenceDependencies(DiagnosticEngine &diags, SourceFile *SF,
     out <<  mangleTypeAsContext(i->first.first);
     out << "\"\n";
   }
+}
 
-  out << "depends-dynamic-lookup:\n";
-  for (auto &entry : sortedByName(tracker->getDynamicLookupNames())) {
+static void writeExternalDynamicLookupDependencies(
+    const ArrayRef<std::pair<DeclBaseName, bool>> sortedDynamicLookupNames,
+    llvm::raw_fd_ostream &out) {
+  out << dynamicLookupDependencies << ":\n";
+  for (auto &entry : sortedDynamicLookupNames) {
     assert(!entry.first.empty());
     out << "- ";
     if (!entry.second)
       out << "!private ";
     out << "\"" << escape(entry.first) << "\"\n";
   }
+}
 
-  out << "depends-external:\n";
-  for (auto &entry : reversePathSortedFilenames(depTracker.getDependencies())) {
+static void writeExternalDependencies(const DependencyTracker &depTracker,
+                                      llvm::raw_fd_ostream &out) {
+  out << externalDependencies << ":\n";
+  for (const auto &entry :
+       reversePathSortedFilenames(depTracker.getDependencies()))
     out << "- \"" << llvm::yaml::escape(entry) << "\"\n";
+}
+
+static void writeInterfaceHash(SourceFile *SF, llvm::raw_fd_ostream &out) {
+  llvm::SmallString<32> interfaceHashVal;
+  SF->getInterfaceHash(interfaceHashVal);
+  out << interfaceHash << ": \"" << interfaceHashVal << "\"\n";
+}
+
+static void writeProviders(const SourceFile *SF, llvm::raw_fd_ostream &out) {
+
+  llvm::MapVector<const NominalTypeDecl *, bool> extendedNominals;
+  llvm::SmallVector<const FuncDecl *, 8> memberOperatorDecls;
+  llvm::SmallVector<const ExtensionDecl *, 8> extensionsWithJustMembers;
+
+  writeTopLevelProviders(SF->Decls, out, extendedNominals, memberOperatorDecls,
+                         extensionsWithJustMembers);
+
+  writeNominalProviders(extendedNominals, out);
+  writeMemberProviders(extendedNominals, extensionsWithJustMembers, out);
+
+  if (SF->getASTContext().LangOpts.EnableObjCInterop)
+    writeDynamicLookupProviders(SF, out);
+}
+
+static void writeDependencies(const SourceFile *SF,
+                              const DependencyTracker &depTracker,
+                              llvm::raw_fd_ostream &out) {
+  const ReferencedNameTracker *const tracker = SF->getReferencedNameTracker();
+  assert(tracker && "Cannot emit reference dependencies without a tracker");
+
+  auto sortedByName = [](const llvm::DenseMap<DeclBaseName, bool> map)
```

Comment:
Yikes, this should definitely have taken the map by reference! Something to fix in follow-up, I guess.

Reply:
Yes, in the follow up. :)